larsim/Utils/TruthMatchUtils.cc

```diff
@@ -0,0 +1,178 @@
+    if (0 == idToEDepMap.size())
+        return kNoG4ID;
+
+    return MaxEDepElementInMap(idToEDepMap)->first;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleIDFromTotalTrueEnergy(const std::vector<art::Ptr<recob::Hit> >& pHits,
+    const bool rollupUnsavedIDs)
+{
+    IDToEDepositMap idToEDepMap;
+    for (const art::Ptr<recob::Hit> &pHit : pHits)
+        TruthMatchUtils::FillG4IDToEnergyDepositMap(idToEDepMap, pHit, rollupUnsavedIDs);
+
+    if (0==idToEDepMap.size())
+        return kNoG4ID;
+
+    return MaxEDepElementInMap(idToEDepMap)->first;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleIDFromTotalRecoCharge(const std::vector<art::Ptr<recob::Hit> >& pHits,
+    const bool rollupUnsavedIDs)
+{
+    IDToEDepositMap idToChargeDepMap;
+    for (const art::Ptr<recob::Hit> &pHit : pHits)
+    {
+        const G4ID g4ID(TruthMatchUtils::TrueParticleID(pHit, rollupUnsavedIDs));
+        const EDeposit recoCharge(static_cast<EDeposit>(pHit->Integral()));
+        std::pair<IDToEDepositMap::iterator, const bool> insertPair(idToChargeDepMap.insert(IDToEDepositMap::value_type(g4ID, recoCharge)));
+        if (!(insertPair.second))
+            insertPair.first->second += recoCharge;
+    }
+
+    if (0 == idToChargeDepMap.size())
+        return kNoG4ID;
+
+    return MaxEDepElementInMap(idToChargeDepMap)->first;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleIDFromTotalRecoHits(const std::vector<art::Ptr<recob::Hit> >& pHits,
+    const bool rollupUnsavedIDs)
+{
+    std::map<G4ID, unsigned int> idToHitCountMap;
+    for (const art::Ptr<recob::Hit> &pHit : pHits)
+    {
+        const G4ID g4ID(TruthMatchUtils::TrueParticleID(pHit, rollupUnsavedIDs));
+        std::pair<std::map<G4ID, unsigned int>::iterator, const bool> insertPair(idToHitCountMap.insert(std::map<G4ID, unsigned int>::value_type(g4ID, 1)));
+        if (!(insertPair.second))
+            insertPair.first->second++;
+    }
+
+    if (0 == idToHitCountMap.size())
+        return kNoG4ID;
+
+    std::map<unsigned int, std::vector<G4ID> > hitCountToIDMap;
+    for (std::map<G4ID, unsigned int>::const_iterator mapIt = idToHitCountMap.begin(); mapIt != idToHitCountMap.end(); ++mapIt)
+    {
+        const G4ID g4ID(mapIt->first);
+        const unsigned int hitCount(mapIt->second);
+        std::pair<std::map<unsigned int, std::vector<G4ID> >::iterator, const bool> insertPair(
+            hitCountToIDMap.insert(std::map<unsigned int, std::vector<G4ID> >::value_type(hitCount, std::vector<G4ID>(1, g4ID))));
```

Comment:
Recommend more concise form:
```.cpp
for (auto const& [g4ID, hitCount] : idToHitCountMap) {
  auto const [iterator, inserted] = hitCountToIDMap.try_emplace(hitCount, {g4ID});
  ...
}
```

Reply:
I've picked up using structured bindings for the for loop and the return from `try_emplace`, to try and keep a consistent style throughout.
Besides what I said above about the structured bindings above, I do have one other reservation; I'm not a massive fan of auto, as I think it breaks one of c++'s main advantages (strong typing).  I'm probably in the minority here though...

Reply:
There are many opinions about `auto`.  However, C++ does not lose strong-typing because of it--the compiler simply calculates what the type is...and that type cannot change for a given name.  It makes sense to use `auto` in small scopes; however once the scopes get too large, then it can be difficult to understand what it refers to.

Reply:
Sorry, I was using terminology incorrectly.  I was basically trying to say that auto makes the code harder to follow and I usually shy away from it.  With structured bindings, it seems like my hand is forced.